feat(Algebra/Order): add Archimedean class

wwylele opened the pull request:
This is the first of PRs for Hahn embedding theorem #25140

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

wwylele commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,425 @@
+in its own class. We denote it as the zero class. -/
+@[to_additive "Following the definition of archimedean classes, the identity element is always
+in its own class. We denote it as the zero class."]
+instance instZero : Zero (mulArchimedeanClass M) := ⟨mk 1⟩
```
I argued with myself on this one a little bit, and eventually decided to go with 0 even for multiplicative case. Reasons are
 - There are more infrastructure in mathlib for zero than for one (`NeZero`, `Finsupp` etc.)
 - Archimedean classes doesn't carry additive/multiplicative structures, so it doesn't seem to matter whether to call this zero or one

eric-wieser replied:
If there's no additive or multiplicative structure, I think you should just use `Top`

wwylele commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,425 @@
+  )
+
+@[to_additive]
+theorem le (A B : mulArchimedeanClass M) : A ≤ B ↔ |B.out|ₘ ≤ |A.out|ₘ := by rfl
```
I don't think this is true. When  `a` and `b` belongs to the same class (meaning we should get `x ≤ y` because `x = y`), `|b|ₘ ≤ |a|ₘ` can be false as the relation between `|a|ₘ` and `|b|ₘ` is just arbitrary.

The `mk_lt_mk` lemma below is probably a better characterization (and then le is lt-or-eq). I have these `.out` lemmas here just so there is a starting point to work with.

eric-wieser replied:
Indeed `mk_lt_mk` looks good; so I'd propose defining this as
```
Quotient.liftOn₂' x y (fun a b => ∃ n, |a|ₘ ≤ |b|ₘ ^ n)
```

wwylele replied:
I'd like to keep `LinearOrder.lift'` to avoid writing all the LinearOrder boiler plate... In the latest change, I added a note to recommend using `mk_lt_mk` and `mk_le_mk`. Is this good enough?

YaelDillies replied:
Aren't we simply missing some API about quotienting linear orders? Something like "If I quotient a linear order by an equivalence relation so that all equivalence classes are order-connected, then the quotient is a linear order too". That sounds like a nice statement that you could quickly prove in a separate PR and easily use here.

YaelDillies replied:
> In the current definition, there is an absolute value in the middle that complicates things a little bit

Oh sorry, I said nonsense: It is true that you are quotienting out the linear order `M`, but not in a way that preserves its order.

Instead, things look like this: You have the total preorder on `M` induced by `(|·|ₘ)` and you identify `a b : M` iff there exist `m n : ℕ` such that `|b|ₘ ≤ |a|ₘ`.
* By default, this gives you a preorder.
* You can upgrade this to a partial order by taking the antisymmetrisation of it.
* You can then upgrade it to a linear order since it has an order embedding into the total preorder on `M` induced by `(|·|ₘ)`.

Using additive notation, here is a second, more principled way to approach this:
1. , then one can define `QuotByAbs M` as the quotient of `M` by the relation `(|·| = |·|)`. `QuotByAbs M` is a canonically linearly ordered monoid.
2. If `M` is a canonically linearly ordered monoid, then one can define `ArchimQuot M` as the antisymmetrisation of the relation `(∃ n : ℕ, · ≤ n • ·)` on `M`. This antisymmetrisation is automatically a linear order.

Then you can define `ArchimedeanClasses M := ArchimQuot (QuotByAbs M)`.

I claim that you were confused precisely because there are two quotients in the above :)

> Perhaps we can introduce a `Preorder` "archimedean LE" `≲`, define the equivalence as `a ≲ b ∧ b ≲ a` and take quotient of the `Preorder` over the equivalence to make it a `PartialOrder` (not sure if this is a well-defined concept)

This is in `Order.Antisymmetrization` 😉

wwylele replied:
Aha! `Antisymmetrization` was another thing I was searching for but didn't know how to spell. I'll see if I can refactor this using it.

I haven't figured out how `QuotByAbs` described here is a well-defined monoid (for example in M = R, (-1)+1 = 0 but their quotient by abs |-1| + |1| ≠ |0|). But I may not need this. `Antisymmetrization` alone makes me feel comfortable to rework this.

YaelDillies replied:
Hmm, I guess I had `|a * b| = |a| * |b|` in mind, but that requires a ring. So yes you're perfectly right

wwylele commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,435 @@
+theorem mk_eq_top_iff {a : M} : mk a = ⊤ ↔ a = 1 := by
+  exact mk_eq_mk_one_iff
+
+@[to_additive]
+theorem ne_top_of_lt {A B : MulArchimedeanClass M} (h : A < B) : A ≠ ⊤ :=
+  lt_of_lt_of_le h le_top |>.ne
+
```
This must have been a leftover after refactoring. I can remove it

wwylele commented:
Thanks for the review! While addressing the comments, I realized the `m ≠ 0` and `n ≠ 0` conditions aren't necessary in the definition of Archimedean equivalence (If `m` satisfying the inequality, them `m' = m + 1  ≠ 0` obviously also satisfies it). This brings the definition closer to the definition of Archimedean group, and I further simplified some proofs with it.

wwylele commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,419 @@
+over archimedean classes."]
+noncomputable
+def orderHom (f : F) : MulArchimedeanClass M →o MulArchimedeanClass N where
+  toFun := fun a ↦ mk (f a.out)
```
I will try making it work.

eric-wieser commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,393 @@
+    simp_rw [← map_mabs, ← map_pow]
+    exact ⟨n, OrderHomClass.monotone f hn⟩
+
+@[to_additive]
+theorem map_mk (f : F) (a : M) : mk (f a) = orderHom f (mk a) := rfl
```
I think this should be
```suggestion
@[to_additive (attr := simp)]
theorem orderHom_mk (f : F) (a : M) : orderHom f (mk a) = mk (f a) := rfl
```

eric-wieser replied:
I think it's still fine to mark it simp, assuming `Quotient.out_eq` has the same treatment.

wwylele replied:
Alright, I added it back

wwylele commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,393 @@
+over archimedean classes."]
+noncomputable
+def orderHom (f : F) : MulArchimedeanClass M →o MulArchimedeanClass N where
+  toFun := Quotient.map f (fun a b h ↦ by
+    obtain ⟨⟨m, hm⟩, ⟨n, hn⟩⟩ := h
+    refine ⟨⟨m, ?_⟩, ⟨n, ?_⟩⟩ <;> rw [← map_mabs, ← map_mabs, ← map_pow]
+    · exact OrderHomClass.monotone f hm
+    · exact OrderHomClass.monotone f hn
+  )
```
:exploding_head: didn't know this was possible

wwylele replied:
actually, this didn't compile for me
```
toFun := Quotient.map f fun | a b ⟨⟨m, hm⟩, ⟨n, hn⟩⟩ => by
```
but this did instead
```
toFun := Quotient.map f fun a b ⟨⟨m, hm⟩, ⟨n, hn⟩⟩ => by
```

eric-wieser replied:
Ah, I think the syntax I was trying needs `,`s, but your version is better anyway.

eric-wieser commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,393 @@
+    · exact OrderHomClass.monotone f hn
+  )
+  monotone' a b h := by
+    rw [← out_eq a, ← out_eq b] at h ⊢
```
I'd argue this would be better as
```suggestion
    induction a using ind with | mk a
    induction b using ind with | mk b
```
after adding
```lean-4
/-- An induction principle for `MulArchimedeanClass`. -/
@[elab_as_elim]
theorem ind {motive : MulArchimedeanClass M → Prop} (mk : ∀ a, motive (.mk a)) : ∀ x, motive x := Quotient.ind mk
```
I'd encourage you to use this style as much as possible, it may simplify things.

wwylele commented:
Updated with the suggestion of using antisymmetrization, now no theorems uses `out` any more (except those with out in the statement, in case I need them in the future).

I noticed that https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Antisymmetrization.html#instLinearOrderAntisymmetrizationLeOfDecidableLEOfDecidableLTOfIsTotal only accepts `≤` relations. Well technically it can be generalized to accepting arbitrary relation, it brings in more questions, such as whether `<` should be derived or provided. In the end, I decided to create a type synonym to equip it with `≤` and `<`, and it looks good and potentially useful for other things (e.g. it can already have a `orderHom` on its own).

wwylele commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,436 @@
+theorem mk_mabs (a : M) : mk |a|ₘ = mk a :=
+  eq.mpr ⟨⟨1, by simp⟩, ⟨1, by simp⟩⟩
+
+@[to_additive (attr := simp)]
+theorem mk_eq_mk_one_iff {a : M} : mk a = mk 1 ↔ a = 1 := by
+  constructor
+  · intro h
+    obtain ⟨_, ⟨_, hm⟩⟩ := eq.mp h
+    simpa using hm
+  · intro h
+    rw [h]
```
`mk 1 = ⊤` is `top_eq` below in the other direction. I can change the direction and make it simp. Meanwhile, `mk_eq_mk_one_iff` (also the `top` version `mk_eq_top_iff`) is important for the `→` direction, which states only `(1 : M)` can make the top element.

YaelDillies replied:
Sure! All I'm saying is that `mk 1` should always be spelled `⊤` )

wwylele commented on Mathlib/Algebra/Order/Group/Unbundled/Abs.lean:
```diff
@@ -213,6 +213,11 @@ variable [Group α] [LinearOrder α] {a b : α}
 
 @[to_additive] lemma lt_of_mabs_lt : |a|ₘ < b → a < b := (le_mabs_self _).trans_lt
 
+@[to_additive] lemma map_mabs {β F : Type*} [Group β] [LinearOrder β] [FunLike F α β]
+    [OrderHomClass F α β] [MonoidHomClass F α β] (f : F) (a : α) :
+    f |a|ₘ = |f a|ₘ := by
```
Did a quick search and indeed a lot of other lemma in similar form (e.g. map_pow) is tagged simp, so I'll tag this as well

wwylele commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,432 @@
+@[to_additive "An ordered group homomorphism can be lifted to an `OrderHom`
+over archimedean classes."]
+noncomputable
+def orderHom (f : F) : MulArchimedeanClass M →o MulArchimedeanClass N :=
```
In general I haven't quite understood when to take a "Class" param and when to take a struct param, as there seems to examples for both in the library. Is the idea that
 - if it is a theorem, take class
 - if it is a def of a new structure, take struct?

btw, this would need to take a `OrderMonoidHom`.

YaelDillies replied:
> if it is a theorem, take class\nif it is a def of a new structure, take struct?

Yes, exactly

wwylele commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,432 @@
+noncomputable
+instance instLinearOrder : LinearOrder (MulArchimedeanClass M) := by
+  classical
+  exact instLinearOrderAntisymmetrizationLeOfDecidableLEOfDecidableLTOfIsTotal
```
a direct `infer_instance` didn't work for me. I found if I unfold `MulArchimedeanClass` then it works with `infer_instance`. I guess that's fine?

YaelDillies commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,436 @@
+def out (A : MulArchimedeanClass M) : M := (Quotient.out A).val
+
+@[to_additive (attr := simp)]
+theorem out_eq (A : MulArchimedeanClass M) : mk A.out = A := Quotient.out_eq' A
```
Doesn't hugely matter for now, but yes it would be nice if you could fix the names in another PR

YaelDillies commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,427 @@
+theorem mk_eq_mk_one_iff {a : M} : mk a = mk 1 ↔ a = 1 := by
+  constructor
+  · intro h
+    obtain ⟨_, ⟨_, hm⟩⟩ := eq.mp h
```
Pointy brackets right-associate
```suggestion
    obtain ⟨_, _, hm⟩ := eq.mp h
```

wwylele commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,427 @@
+  eq.mpr ⟨⟨1, by simp⟩, ⟨1, by simp⟩⟩
+
+@[to_additive]
+theorem mk_eq_mk_one_iff {a : M} : mk a = mk 1 ↔ a = 1 := by
```
Ah, I didn't get it but I think I understand now. I'll just remove all theorems with `mk 1`

YaelDillies commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,418 @@
+  exact (pow_two |b|ₘ ▸ (h 2).le).trans (mabs_mul' a b)
+
+@[to_additive]
+theorem mk_right_eq_mk_mul (h : mk b < mk a) : mk b = mk (a * b) :=
```
```suggestion
theorem mk_mul_eq_mk_right (h : mk b < mk a) : mk (a * b) = mk b :=
```
Same above

YaelDillies reviewed (approved):
Thanks!

maintainer delegate

wwylele commented:
Did a small clean up following similar suggestions from #25168

wwylele commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,410 @@
+
+-/
+
+section Pre
```
Ah, I think I stubbed that name and forgot to change... let me change that

jcommelin commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,410 @@
+# Archimedean classes of a linearly ordered group
+
+This file defines archimedean classes of a given linearly ordered group. Archimedean classes
+measure to what extent the group fails to be Archimedean. For additive group, elements `a` and `b`.
```
```suggestion
measure to what extent the group fails to be Archimedean. For additive group, elements `a` and `b`
```

jcommelin commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,410 @@
+
+namespace MulArchimedeanOrder
+
+/-- Create a `MulArchimedeanOrder` element from the underlying type. -/
+@[to_additive "Create a `ArchimedeanOrder` element from the underlying type."]
+def of (a : M) : MulArchimedeanOrder M := a
+
+/-- Retrieve the underlying value from a `MulArchimedeanOrder` element. -/
+@[to_additive "Retrieve the underlying value from a `ArchimedeanOrder` element."]
+def val (a : MulArchimedeanOrder M) : M := a
```
For type synonyms, we usually set up an equiv between the old and new type. See for example https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Group/TypeTags/Basic.html#Multiplicative.ofAdd

jcommelin commented on Mathlib/Algebra/Order/Archimedean/Class.lean:
```diff
@@ -0,0 +1,410 @@
+@[to_additive]
+theorem mk_lt_mk : mk a < mk b ↔ ∀ n, |b|ₘ ^ n < |a|ₘ := .rfl
+
+/-- 1 is in its own class, which is also the largest class. -/
+@[to_additive "0 is in its own class, which is also the largest class."]
```
```suggestion
/-- 1 is in its own class (see `MulArchimedeanClass.mk_eq_top_iff`), which is also the largest class. -/
@[to_additive "0 is in its own class (see `ArchimedeanClass.mk_eq_top_iff`), which is also the largest class."]
```

jcommelin reviewed (commented):
Thanks :tada:

bors merge
